src/main/webapp/app/complaints/complaint-interactions.component.ts

```diff
@@ -113,36 +110,34 @@ export class ComplaintInteractionsComponent implements OnInit {
      */
     get isTimeOfComplaintValid(): boolean {
         if (this.isExamMode) {
-            if (this.testRun) {
-                return !!this.result && !!this.result.completionDate;
+            if (!!this.result && !!this.result.completionDate) {
+                return this.testRun || this.isWithinStudentReviewPeriod();
             }
-            return this.isWithinStudentReviewPeriod() && !!this.result && !!this.result.completionDate;
         } else if (this.result && this.result.completionDate) {
-            const resultCompletionDate = moment(this.result.completionDate!);
-            if (!this.exercise.assessmentDueDate || resultCompletionDate.isAfter(this.exercise.assessmentDueDate)) {
-                return resultCompletionDate.isAfter(moment().subtract(this.course?.maxComplaintTimeDays, 'day'));
-            }
-            return moment(this.exercise.assessmentDueDate).isAfter(moment().subtract(this.course?.maxComplaintTimeDays, 'day'));
-        } else {
-            return false;
+            return this.isComplaintWithinTimeBoundaries();
         }
+        return false;
     }
 
     /**
      * Analogous to isTimeOfComplaintValid but exams cannot have more feedback requests.
      */
     get isTimeOfFeedbackRequestValid(): boolean {
-        if (this.isExamMode) {
-            return false;
-        } else if (this.result && this.result.completionDate) {
-            const resultCompletionDate = moment(this.result.completionDate!);
-            if (!this.exercise.assessmentDueDate || resultCompletionDate.isAfter(this.exercise.assessmentDueDate)) {
-                return resultCompletionDate.isAfter(moment().subtract(this.course?.maxRequestMoreFeedbackTimeDays, 'day'));
-            }
-            return moment(this.exercise.assessmentDueDate).isAfter(moment().subtract(this.course?.maxRequestMoreFeedbackTimeDays, 'day'));
-        } else {
-            return false;
+        if (!this.isExamMode && this.result && this.result.completionDate) {
+            return this.isComplaintWithinTimeBoundaries();
+        }
+        return false;
+    }
+
+    /**
+     * Checks if a complaint (either actual complaint or more feedback request) can be filed
+     */
+    isComplaintWithinTimeBoundaries(): boolean {
+        const resultCompletionDate = moment(this.result.completionDate!);
+        if (!this.exercise.assessmentDueDate || resultCompletionDate.isAfter(this.exercise.assessmentDueDate)) {
```

Comment:
Can we ensure that the exercise is not null? Might add a `this.exercise?.assessmentDueDate` for safety reasons

Reply:
I'd say so. this.exercise is not optional on line 20.

Reply:
Yeah if we add an optional `?` to the Imports we would expect exercise to eventually be null, but now it could still be null even though we don't want it to be, would not throw an error until line 137 I assume. But as you just moved this code it should be fine